File: src/main/java/edu/cornell/cs/apl/viaduct/errors/IllegalCharacterError.java
package edu.cornell.cs.apl.viaduct.errors;

import edu.cornell.cs.apl.viaduct.imp.parsing.SourceRange;
import java.io.PrintStream;

/** Thrown when the lexer encounters an illegal character. */
public class IllegalCharacterError extends CompilationError {
  private final SourceRange location;

  public IllegalCharacterError(SourceRange location) {
    this.location = location;
  }

  @Override
  protected String getCategory() {
    return "Parse Error";
  }

  @Override
  protected String getSource() {
    return location.getSourcePath();
  }

  @Override
  public void print(PrintStream output) {
    super.print(output);

    output.println("I ran into a character I did not expect:");

    output.println();
    location.showInSource(output);
  }
}

File: src/main/kotlin/edu/cornell/cs/apl/viaduct/errorskotlin/IllegalCharacterError.kt
package edu.cornell.cs.apl.viaduct.errorskotlin

import edu.cornell.cs.apl.prettyprinting.Document
import edu.cornell.cs.apl.viaduct.syntax.SourceLocation

/**
 * Thrown when the lexer encounters an illegal character.
 *
 * @param location Source location of the illegal character.
 */
class IllegalCharacterError(private val location: SourceLocation) : CompilationError() {
    override val category: String
        get() = "Parse Error"

    override val source: String
        get() = location.sourcePath

    override val description: Document
        get() =
            Document("I ran into a character I did not expect:").withSource(location)
}
